monash-human-power/MHP_DAS
# Crash Detection

## MPU Sensor
The MPU6050 is used to measure and calculate the state of the bike. Then the wireless module publishes the information to the MQTT broker under the /v3/wireless_module/3/mpu_data topic for acceleration, rotation and gyroscopic data, and /v3/wireless_module/3/crashed_state for crashed state.

## Setup
1. Follow the Getting Started instructions [here](https://github.com/monash-human-power/data-acquisition-system/tree/master/wireless_modules) 
2. Once you have created a local version of the config.py file you can navigate to the ```back_module``` folder.
3. Execute the upload script with 
```bash
# change this to the serial port of the ESP32
./upload.sh --port /dev/ttyUSB0
```
``` bash
# if this is taking a while in your shell, switch shells or alternatively manually upload these files
(
    "main.py" 
    "../config.py" 
    "../wireless_module.py"
    "../boot.py"
    "../mqtt_client.py"
    "../battery_reader.py"
    "../sensors/sensor_base.py"
    "../sensors/co2_sensor.py"
    "../sensors/dht_sensor.py"
    "../sensors/gps_sensor.py"
    "../sensors/mpu_sensor.py"
    "../sensors/reed_sensor.py"
    "../status_led.py"
    "../libraries/abc.py"
    "../libraries/MQ135/mq135.py"
    "../libraries/micropyGPS/micropyGPS.py"
    "../libraries/MPU6050-ESP8266-MicroPython/mpu6050.py"
) 
# using 
ampy -p /dev/ttyUSB0 -b 115200 put <FILE>
```
4. Verify that the file system is uploaded correctly using
```bash
# change /dev/ttyUSB0 to the serial port of the ESP32
ampy -p /dev/ttyUSB0 ls
```
5. Install and setup [Mosquitto](https://mosquitto.org/download/) and make sure its running

## Steps to Test

1. Switch on the primary display
3. Create a mobile hotspot with the previously used ESSID and password
5. Verify that the wireless module, primary display and subscribed devices are all connected to the hotspot
7. Interact with the wireless module through your computer (see [this](https://github.com/monash-human-power/data-acquisition-system/tree/master/wireless_modules/middle_module#steps-to-test))


## Current Tuning Method
It is a good idea to calibrate the sensors on before using it. Notably aim to calibrate the sensor if: the back wireless module is accidently dropped, or: after real crash has been detected.
In the current implementation of the accelerometer, you will need to manually calibrate the sensor if the readings have drifted. Parameters are stored as class attributes of the ```MpuSensor``` within ```mpu_sensor.py```. These tunable values include:
| Parameter| Description |
| --- | ----------- |
| ```xyz_calib_offset``` | Is used to ensure axis readings are zero while perpendicular to the direction of gravity. To tune this parameter. </br>1. Modify the mpu_sensor.py file by adding ```print(accel_values) ``` in line 80 </br>2. Lay the accelerometer with z-axis facing up, and ensure that the accelerometer in on a level surface with a leveler and that there is as little vibration as possible </br>3. Test the device and read the values of x and z. </br> 4. Pick values ```a,b``` for the other two axis x,y such that ```x+a≈0``` and ```y+b≈0```  <br> 5. Verify these values by reorienting the accelerometer with other axis facing directly up. </br> 5. Update the class attribute and reflash the wireless module again before moving on to the next parameter |
| ```xyz_calib_factor``` | This determines how sensitive the accelerometer is. Once xyz_calib_offset it tuning sufficiently, </br>1. Place the accelerometer on a level surface with z axis facing up, find the value of ```a``` such that ```z*a≈1``` </br> 2. Repeat this by reorienting the accelerometer with the calibration axis facing up and solving for ```a``` once again for all other axis readings </br>3. Update the parameter and reflash the wireless module|
| ```gyro_calib_offset``` | This parameter helps to offset the initial error in the gyroscope. Once xyz_calib_factor is complete </br> 1. Place MPU on a flat surface with as little vibration as possible, modify the ```mpu_sensor.py``` code to include ```print(gyro_values)``` instead on ```line 80```  and reflash the wireless module</br> 2. Pick values for ```gyro_calib_offset``` such that the printed values are as close to 0 as possible while at rest </br>3. Reflash the wireless module with the new values.</br> A method will be added to calibrate this automatically soon.|
| complemenatary ratio | The complementary ratio is not store as a class attribute. The tuning is left to the judgement of the user. See [notion](https://www.notion.so/monashhumanpower/Wireless-module-crash-detection-5295c2b1b84e44f980fb0b6abb87bd92) for approach. The actual values are the hardcoded values found in ```line 74``` and ```line 75``` in ```mpu_sensor.py``` |

Currently there is no way to tune gyroscope scaling without the use of specialized testing equipment.
